### ISSUE TITLE
[Suggestion] Partial workaround for multi-row dragging problem

### ISSUE THREAD

**Issue by sonofevil** [MrOtherGuy/firefox-csshacks#161]
I'm working on a Waterfox port of the multi-row tab stylesheet, and I found a partial workaround for the problem of not being able to drag tabs when they are on multiple rows.

The idea is that once tabs exceed the width of the first line, all unpinned tabs are moved to the next line together. Since dragging for pinned and unpinned tabs is handled separately, this assures that dragging works for as long as possible. This will be most useful for people who have a lot of pinned tabs.

The way this works is that all pinned and unpinned tabs each are assigned a certain percentage of the full width of the tab bar, allowing for mathematical predictability. Then a fairly clunky selector can account for all possible line-breaking cases and insert a line break in the right place once those conditions are met.

Only the bottom part of the stylesheet matters:
https://github.com/sonofevil/waterfox-csshacks/blob/patch-1/Waterfox/chrome/multi-row_tabs.css

I haven't tested this with Firefox, but it should be easy to port it back.

**Comment by MrOtherGuy** [comment]
Well, I tested that with Firefox (which required some heavy-ish changes) and maybe it doesn't work like it does on Waterfox. At least, I don't see how the behavior is any better than what the multi_row_tabs.css has. The only difference I can tell is that pinned tabs are on separate row - that is by itself kinda neat and maybe someone would prefer that (and can be done with pretty simple css addition), but I don't see how it makes tab dragging any better.

**Comment by sonofevil** [comment]
Which part exactly didn't work, if you don't mind me asking? Did it fail to keep both kinds of tabs on the same line when they didn't exceed the width together? Or something else?

It makes tab dragging better insofar as it delays the point at which it breaks.

**Comment by MrOtherGuy** [comment]
Sorry, it's been a while but I've now managed to give this a bit more thought.

It didn't appear to have any noticeable effect because I didn't try it with enough pinned tabs for it to really matter.

Also a thing I didn't understand previously was that you intended to make the tabs show on one line if they seem to fit.  That would be nice to have indeed, but I really don't want to do that IF it requires us to set width for tabs - and it looks to me to be a necessity.

For now, I have made a [patch file](https://github.com/MrOtherGuy/firefox-csshacks/blob/master/chrome/multi-row_tabs_separate_pinned_row_patch.css) to multi-row_tabs.css that just separates pinned tabs to separate row(s). If we can figure out a clean way to make non-pinned tabs show on the same row should they fit then it would be nice to add that.

**Comment by sonofevil** [comment]
Alright, here's a kinda hacky workaround for most dragging situations:

`#tabbrowser-tabs[movingtab="true"] arrowscrollbox {
  text-align: center !important;
}

#tabbrowser-tabs[movingtab="true"]  .tabbrowser-tab {
  width: 2% !important;
  min-width: 2% !important;
}

#tabbrowser-tabs[movingtab="true"]  .tabbrowser-tab .tab-close-button {
  display: none;
}

#tabbrowser-tabs[movingtab="true"]  .tabbrowser-tab .tab-content {
  padding: unset !important;
}`

This simply shrinks all tabs down to 2% of the box width, meaning dragging will work with up to 50 tabs, or up to 50 pinned and 50 unpinned tabs if you use my previous line-break hack.
Disadvantages: 
- The dragged tab will initially change its position and no longer be beneath the mouse.
- You can no longer read the labels of unpinned tabs while dragging.

**Comment by sonofevil** [comment]
There is potential in the movingtab=true selector, I'm trying to make it so that when a tab is dragged, the layout resets to Firefox default. Only problem is that the scrollbox element is inside a shadow root, and the movingtab=true property is outside of it. I have no idea how to circumvent this.

**Comment by MrOtherGuy** [comment]
I think I tested doing that sort of "revert to normal" while dragging tabs at some point. It just wasn't all that good because the dragged tab might end up anywhere in the tab strip. But I would certainly be interested if you can make it work somehow.

**Comment by sonofevil** [comment]
I may have found a different solution:

`.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(5) {
    margin-right: calc(1 * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(5) + .tabbrowser-tab:not([pinned],[first-visible-unpinned-tab]) {
  margin-left: calc((-1) * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(6) {
    margin-right: calc(2 * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(6) + .tabbrowser-tab:not([pinned],[first-visible-unpinned-tab]) {
  margin-left: calc((-2) * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(7) {
    margin-right: calc(3 * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(7) + .tabbrowser-tab:not([pinned],[first-visible-unpinned-tab]) {
  margin-left: calc((-3) * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(8) {
    margin-right: calc(4 * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(8) + .tabbrowser-tab:not([pinned],[first-visible-unpinned-tab]) {
  margin-left: calc((-4) * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(9) {
    margin-right: calc(5 * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(9) + .tabbrowser-tab:not([pinned],[first-visible-unpinned-tab]) {
  margin-left: calc((-5) * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(10) {
    margin-right: calc(6 * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(10) + .tabbrowser-tab:not([pinned],[first-visible-unpinned-tab]) {
  margin-left: calc((-6) * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(11) {
    margin-right: calc(7 * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(11) + .tabbrowser-tab:not([pinned],[first-visible-unpinned-tab]) {
  margin-left: calc((-7) * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(12) {
    margin-right: calc(8 * var(--multirow-tab-min-width)) !important;
}

.tabbrowser-tab[first-visible-unpinned-tab]:nth-last-child(12) + .tabbrowser-tab:not([pinned],[first-visible-unpinned-tab]) {
  margin-left: calc((-8) * var(--multirow-tab-min-width)) !important;
}

... and so forth`

Of course this is ugly and clunky. But it works. The unpinned tabs all break to the next line together when necessary. This is achieved by making the right margin of the first unpinned tab the same as the minimum length of all the tabs next to it, and making the other tabs "ignore" that margin by inverting it for the first one. That way the flexbox treats all unpinned tabs as if they were one wide tab. Unlike my earlier suggestion, this solution is agnostic wrt the width of the window and the width of the tabs.

Now if anyone has an idea how to optimize this, that would be great. I tried nesting, but that doesn't seem to work in the Firefox GUI css.

**Comment by MrOtherGuy** [comment]
That sure looks ugly, though you could make it bit tidier by removing all those `:not([pinned],[first-visible-unpinned-tab])` parts, since the next tab after one that has [first-visible-unpinned-tab] can't possibly be pinned or first-visible-unpinned-tab itself.

However, I fail to see how that changes anything. If I add that to *just* multi-row_tabs.css then the behavior is pretty much exactly the same as normally. But if I also include that _separate_pinned_tabs_row.css file then this doesn't seem to have any effect at all and pinned tabs still have their completely separate row.

**Comment by sonofevil** [comment]
Yeah it's pointless together with the `_separate_pinned_tabs_row.css` because it's an alternative to it. A good way to visualize what this does is to imagine a `_separate_pinned_tabs_row.css` that only activates under certain conditions. The condition that activates it is when all tabs no longer fit into one row. When all tabs would fit into one row, `_separate_pinned_tabs_row.css` isn't needed because dragging still works.

Are you sure you tested the relevant cases? 

> It didn't appear to have any noticeable effect because I didn't try it with enough pinned tabs for it to really matter.
> 
> Also a thing I didn't understand previously was that you intended to make the tabs show on one line if they seem to fit.

**Comment by sonofevil** [comment]
Case A (inactive): 
![Screenshot_20221213_112244](https://user-images.githubusercontent.com/19374890/207292741-1d058010-f887-45a2-a3ad-9b146b4d8f0e.png)

Case B (active):
![Screenshot_20221213_112309](https://user-images.githubusercontent.com/19374890/207292804-c7c422d3-d1d2-4572-9ad7-8fa469e42f72.png)

**Comment by MrOtherGuy** [comment]
Okay, now I see what you mean. I thought it depended on number of pinned tabs, but number of *unpinned* tabs seems equally (or even more?) important here. I experienced a scenario where unpinned tabs still wrapped to another row one by one just like how multi-row_tabs normally works. But opening more unpinned tabs made it eventually wrap them all to separate line.

So, what I see happening is that I can have all my pinned tabs on top row followed by few unpinned tabs in the top row, and then some unpinned tabs on another row. Then if I open a new (unpinned) tab, it may cause *all* unpinned tabs to reflow to second line.

**Comment by sonofevil** [comment]
I see, that sounds like a glitch caused by the formula calculating the margin. In my browser it works perfectly, but it seems that in your browser, the margin is too short for some reason I can't troubleshoot myself. So you need to have to open a few more tabs before the margin gets long enough to cause the first unpinned tab to wrap.

To clarify, this is how the margin works:
![Screenshot_20221213_115401](https://user-images.githubusercontent.com/19374890/207299133-54101e32-495b-4aad-9c05-b19164110b3a.png)

**Comment by sonofevil** [comment]
The idea was that one way to achieve the desired behavior would be to nest unpinned tabs together into a single element in the html, and the container would basically act like a single big tab that cannot be split across rows. It occured to me that since tabs have a minimum width, it is possible to emulate that behavior by using the first unpinned tab's right margin as a quasi-container. The first unpinned tab then effectively has the same width as all unpinned tabs taken together, and thus wraps at the correct times, pushing the other tabs with it. It's really not that complicated.

I don't think it depends on any specific setup. Literally the only thing it depends on is the effective minimum width of unpinned tabs. That's why it uses that variable.

There is another possible explanation for why it doesn't work in your case. The width could be calculated correctly, but the css assumes that the last unpinned tab is the 4th to last element in the html. Your browser's html might be different for some reason, so that more than three elements come after the last tab, so that the css misapprehends the number of unpinned tabs.

This should be easy enough to fix. Pretty sure I can generalize to make it agnostic regarding the number of sibling elements coming after the last tab.

**Comment by MrOtherGuy** [comment]
> The 3 elements coming after the last tab in my browser:
> ![Screenshot_20221213_135933](https://user-images.githubusercontent.com/19374890/207324746-2ebb3e8e-a5e2-4d8f-a8c4-2a6798ba256e.png)

This looks wrong. There should be an additional container element with id `tabbrowser-arrowscrollbox-periphery` immediately after the last `<tab>` element which contains tabs-newtab-button and closing-tabs-spacer.

Are you sure you are testing with up-to-date Firefox?

**Comment by sonofevil** [comment]
Now I am. Can confirm that my latest [snippet](https://github.com/MrOtherGuy/firefox-csshacks/issues/161#issuecomment-1348580664) works there perfectly.

**Comment by MrOtherGuy** [comment]
Yeah, that latest snip seems better, but at least on my system it kinda breaks when further tabs are created. Like, the non-pinned tabs go to the same row as pinned tabs once again. So if you are at limit of what fits a single line then all tabs are in that single line. Good. Open a new tab, they don't fit anymore, so all non-pinned tabs flow to second line. Good again. Then open a couple more tabs, the tabs flow back to filling the same row where pinned tabs are. Not good.

I think this is exactly what I mentioned earlier, like you would need to create a rule for each potential tab for it to work correctly. It's not enough to consider just pinned tabs because you are using `:nth-last-of-type()` so any tabs you open "change what that n is". Maybe it would work better if you used `:nth-of-type()` instead?

**Comment by sonofevil** [comment]
`Open a new tab, they don't fit anymore, so all non-pinned tabs flow to second line. Good again. Then open a couple more tabs, the tabs flow back to filling the same row where pinned tabs are. Not good.`

Ah yeah, I know what you mean. I actually think it's good for them to go back to the first line when they are wrapping anyway, and hence dragging is broken anyway, because it saves vertical space, but they should do so immediately when they fill a row. But the point at which they do this depends on where the rules stop applying. So the behavior ends up seeming somewhat arbitrary when that happens.

**Comment by sonofevil** [comment]
Seems like this addition would achieve the behavior you want:

```
tab[first-visible-unpinned-tab]:nth-last-of-type(15+n)       { margin-right: calc(   14+n  * var(--mrtmw)) !important; }
tab[first-visible-unpinned-tab]:nth-last-of-type(15+n) + tab { margin-left:  calc((-14 +n) * var(--mrtmw)) !important; }
```

**Comment by sonofevil** [comment]
There is definitely a way to achieve the behavior I want using `vw` and `mod()`. Unfortunately `mod()` is extremely new and unsupported.
